jumanji/training/networks/graph_coloring/actor_critic.py

```diff
@@ -0,0 +1,283 @@
+        nodes_leaves = jnp.stack(jtree.tree_leaves(nodes_float), axis=-1)
+        embeddings = hk.Linear(self.model_size, name="node_projection")(nodes_leaves)
+        return embeddings
+
+    def embed_colors(self, colors: chex.Array) -> chex.Array:
+        color_float = colors.astype(jnp.float32)
+        colors_leaves = jnp.stack(jtree.tree_leaves(color_float), axis=-1)
+        embeddings = hk.Linear(self.model_size, name="color_projection")(colors_leaves)
+        return embeddings
+
+    def __call__(self, observation: Observation) -> chex.Array:
+        # observation.action_mask => colors available
+        # observation.colors => nodes that are colored
+        # node embedding check if it's colored or not
+        # cross attention => for each color true if the node is colored with it
+
+        batch_size, num_nodes = observation.colors.shape
+        colors_range = jnp.arange(num_nodes).reshape(1, -1)
+        colors_used = jnp.any(
+            observation.colors[..., jnp.newaxis] == colors_range, axis=1
+        )
+
+        nodes_colored = observation.colors >= 0
+
+        color_embeddings = self.embed_colors(
+            colors_used
+        )  # Shape (batch_size, num_colors, 128)
+        node_embeddings = self.embed_nodes(
+            nodes_colored
+        )  # Shape (batch_size, num_colors, 128)
+
+        mask = self._make_self_attention_mask(
+            observation.adj_matrix
+        )  # Shape (batch_size, transformer_num_heads, num_nodes, num_nodes)
+
+        colors_array = jnp.expand_dims(
+            observation.colors, axis=1
+        )  # Shape (batch_size, 1, num_nodes)
+        color_indices = jnp.arange(observation.colors.shape[1])  # Shape (num_colors,)
+        color_indices = jnp.expand_dims(color_indices, axis=0)  # Shape (1, num_colors)
+        color_indices = jnp.expand_dims(
+            color_indices, axis=-1
+        )  # Shape (1, num_colors, 1)
+
+        colors_cross_nodes_mask = (
+            colors_array == color_indices
+        )  # Shape (batch_size, num_colors, num_nodes)
+
+        # Expand and repeat along the transformer_num_heads axis
+        colors_cross_nodes_mask = jnp.expand_dims(
+            colors_cross_nodes_mask, axis=1
+        )  # Shape (batch_size, 1, num_colors, num_nodes)
+        colors_cross_nodes_mask = jnp.repeat(
+            colors_cross_nodes_mask, repeats=self.transformer_num_heads, axis=1
+        )  # Shape (batch_size, transformer_num_heads, num_colors, num_nodes)
+
+        colors_array = jnp.expand_dims(
+            observation.colors, axis=-1
+        )  # Shape (batch_size, num_nodes, 1)
+        color_indices = jnp.arange(observation.colors.shape[1])  # Shape (num_colors,)
+        color_indices = jnp.expand_dims(color_indices, axis=0)  # Shape (1, num_colors)
+        color_indices = jnp.expand_dims(
+            color_indices, axis=0
+        )  # Shape (1, 1, num_colors)
+
+        nodes_cross_colors_mask = (
```

Comment:
It seems to me here that `nodes_cross_colors_mask` is equal to `colors_cross_nodes_mask`, but they should be a transpose of each other, right?

Reply:
I wouldn't say they are the transpose of each other. 

`colors_cross_nodes_mask` represents for each color, which nodes have that color. It has shape (batch_size, transformer_num_heads, num_colors, num_nodes).

`nodes_cross_colors_mask` represents for each node, which colors it has. It has shape (batch_size, transformer_num_heads, num_nodes, num_colors).

Reply:
Am I missing something or is this not exactly the transpose of one another?
I.e. `colors_cross_nodes_mask[i, j]` is `True` if node `j` has to be colored by color `i`. And `nodes_cross_colors_mask[i, j]` is `True` if node `i` has to be colored by color `j`.

Am I missing something?

Reply:
Yes, you're right. Gotcha!

Reply:
@clement-bonnet,

Any comment on this or should I resolve the issue?